pymeasure/adapters/adapter.py

```diff
@@ -34,44 +37,112 @@ class Adapter:
     This class should only be inherited from.
 
     :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+        received from the instrument. The callable returns the processed
+        string. This argument is deprecated.
     :param kwargs: all other keyword arguments are ignored.
     """
 
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn(
+                "Do not use `Adapter.values`, but `Instrument.values` instead. "
+                "Preprocess_reply is a values parameter.",
+                FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
-        if self.connection is not None:
+        self.close()
+
+    def close(self):
+        """Close the connection."""
+        try:
             self.connection.close()
+        except AttributeError:
+            pass
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
+    def write(self, command, **kwargs):
+        """
+        Write a string command to the instrument.
 
-        :param command: SCPI command string to be sent to the instrument
+        The write_termination string is appended, if defined.
+
+        :param command: Command string to be sent to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", command))
```

Comment:
Why the double parentheses? Is that for the bytestring aspects of the test generator?

Reply:
The first argument is going to be logged, the rest are further parameters like exception, line number etc.
I want to log (for the test generator, as you thought) a tuple of the string indicator and the content itself. With double parentheses I pass a tuple as first argument.

Reply:
> The first argument is going to be logged, the rest are further parameters like exception, line number etc.

Afaik, posargs are for messages to use in string formatting, kwargs for those "further" parameters, so 
`self.log.debug("WRITE:%s", "bla")` would log "WRITE:bla" and perform string formatting only if necessary (for performance reasons)

Reply:
It feels weird to have that extra set of parenthese (the logged tuple) in the log, because of the internal use of the test generator...can we deal with that on the test generator side?

Reply:
The logger will internally aggregate everything to one string, but I want to keep it as a bytes object, lest I have to parse a byte string into a bytes object.
An alternative is to use a flag value for "logging" and use our own or list etc.

Reply:
I'll investigate more

Reply:
> I'll investigate more

Thanks!

> The logger will internally aggregate everything to one string, 

Ah, so the console log message comes out without the tuple parentheses? I made a quick test in the console, and there I still saw the tuple in the message:
```
In [8]: logging.warning(('ba', "bb"))
WARNING:root:('ba', 'bb')
```

Reply:
I exchanged the extra pair of parentheses for "%s", such that normal logging works as expected, and I'll still be able to extract the bytes data for the Generator.

Reply:
That's great! :-)